### PR TITLE
Import and optimization of PEtab models

### PR DESCRIPTION
**Pull request by dweindl**
TODO:
- [x] Mapping of AMICI gradients to cost function gradient
- [x] Mapping of AMICI hessian to cost function Hessian
- [x] Condition-dependent setting of parameter scaling
- [x] For constant parameters assume linear scale
~~- [ ] Setting AMICI ParameterList (currently done w.r.t all parameters)~~ (#91, can be done later)
- [x] Cleanup extra files and redundant code
- [x] Remove redundant simulation code in AmiciObjective
~~- [ ] write tests for petab import~~ ( #97, can be done later)

Update version 0.0.2 -> 0.0.3

### CODE REVIEW

File: pypesto/objective/petab_import.py [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+        if self.output_folder not in sys.path:
+            sys.path.insert(0, self.output_folder)
+
+        # load moduĺe
+        model_module = importlib.import_module(self.petab_problem.name)
+
+        # import model
+        self.model = model_module.getModel()
+
+    def compile_model(self):
```
Comment (dweindl) [inline pypesto/objective/petab_import.py]: Shall we move this one to AMICI? Might be easier to keep it in sync with the code it depends on and may be more reusable.

Comment (yannikschaelte) [reply pypesto/objective/petab_import.py]: definitely an option. why i didn't so far:
1) i did not want to split up the funtcionality into too many places (petab, pypesto.petab_import, pypesto.amici_objective, amici ...)
2) the comiplation-specific part is rather short, then already the objective creation begins

File: pypesto/objective/petab_import.py [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+            sigma_y = np.full(
+                shape=(edata.nt(), edata.nytrue()),
+                fill_value=np.nan)
+
+            # add measurements and sigmas
+            for _, measurement in cur_measurement_df.iterrows():
+                time_ix = timepoints.index(measurement.time)
+                observable_ix = observable_ids.index(
+                    f'observable_{measurement.observableId}')
+                # TODO: measurement file should contain prefix
```
Comment (dweindl) [inline pypesto/objective/petab_import.py]: Should or shouldn't? I think we agreed to leave "observable_" out there.

File: pypesto/objective/petab_import.py [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+        scale_mapping = \
+            petab.core.get_optimization_to_simulation_scale_mapping(
+                parameter_df=self.petab_problem.parameter_df,
+                mapping_par_opt_to_par_sim=parameter_mapping
+            )
+
+        # create objective
+        obj = AmiciObjective(
+            amici_model=self.model, amici_solver=self.solver, edata=edatas,
+            x_ids=par_opt_ids, x_names=par_opt_ids,
```
Comment (dweindl) [inline pypesto/objective/petab_import.py]: Could try getting parameter names from petab files first and only then fall back to IDs. But also fine to leave it for now.

File: pypesto/objective/petab_import.py [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+                # append to dataframe
+                df = df.append(row_sim, ignore_index=True)
+
+        return df
+
+
+def _get_simulation_conditions(condition_df, measurement_df):
+    """
+    Compute the conditions by which to group the measurements, so
+    that for each group an amici.EData can be generated.
```
Comment (dweindl) [inline pypesto/objective/petab_import.py]:
```suggestion
    that for each group an amici.ExpData can be generated.
```

File: doc/releasenotes.rst [hunk 5]
```diff
@@ -5,13 +5,23 @@ Release notes
 0.0 series
 ..........
 
+
+0.0.3 (2019-01-30)
+------------------
+
+* Import amici models and the petab data format automatically using
+  pypesto.PetabImporter.
+* Basic profiling routines.
```
Comment (yannikschaelte) [inline doc/releasenotes.rst]: @paulstapor Could you shortly describe here what exactly exists already (unless "Basic profiling routines" is a good enough description already).

File: pypesto/objective/amici_objective.py [hunk 6]
```diff
@@ -16,31 +17,67 @@ class AmiciObjective(Objective):
...
+            The default is just to assume that optimization and simulation
+            parameters coincide. The default is to assume equality of both.
+
+        mapping_scale_opt_to_scale_sim: optional
+            Mapping of optimization parameter scales to simulation parameter
+            scales. The default is to just use the scales specified in the
+            `amici_model` already.
+
+        preprocess_edata: bool, optional
+            Whether to preprocess the experimental data.
```
[the remark this answers is not in the document]
Comment (dweindl) [reply pypesto/objective/amici_objective.py]: Please document.

File: pypesto/objective/amici_objective.py [hunk 7]
```diff
@@ -196,87 +272,90 @@ def _call_amici(
...
+                        rdata['sllh'],
+                        snllh,
+                        coefficient=-1.0
+                    )
                     # TODO: Compute the full Hessian, and check here
-                    ssnllh -= rdata['FIM']
+                    if sensi_order > 1:
+                        pass
+                    else:
+                        add_sim_hess_to_opt_hess(
```
[the remark this answers is not in the document]
Comment (yannikschaelte) [reply pypesto/objective/amici_objective.py]: yeah, that does not make sense. as of now, will assume we always have the fim.

File: pypesto/objective/amici_objective.py [hunk 8]
```diff
@@ -285,52 +364,145 @@ def _call_amici(
...
                 original_initial_state_sensitivities = \
                     self.amici_model.getInitialStateSensitivities()
 
-            fixed_parameters = copy.deepcopy(
-                list(data.fixedParametersPreequilibration)
-            )
+            # remember original fixed parameters for preequilibration
+            original_fixed_parameters_preequilibration \
+                = data.fixedParametersPreequilibration
+            # unset fixed preequilibration parameters in data (TODO: Why?)
```
[the remark this answers is not in the document]
Comment (yannikschaelte) [reply pypesto/objective/amici_objective.py]: ah ok. Thanks!

File: pypesto/objective/amici_objective.py [hunk 9]
```diff
@@ -69,20 +106,51 @@ def __init__(self, amici_model, amici_solver, edata, max_sensi_order=None,
 
         self.amici_model = amici_model
         self.amici_solver = amici_solver
-        self.dim = len(amici_model.getParameterList())
+
+        # make sure the edatas are a list of edata objects
+        if not isinstance(edata, list):
```
[the remark this answers is not in the document]
Comment (yannikschaelte) [reply pypesto/objective/amici_objective.py]: @dweindl If I change that to ``isinstance(edata, amici.ExpData)``, it complains about ``TypeError: isinstance() arg 2 must be a type or tuple of types``. Do you know how to do this?

[the remark this answers is not in the document]
Comment (dweindl) [reply pypesto/objective/amici_objective.py]: Almost. Use `isinstance(edata, amici.amici.ExpData)`. This is masked by `amici.ExpData` which is a python wrapper of the C++ constructor.

File: pypesto/objective/amici_objective.py [hunk 10]
```diff
@@ -340,36 +512,206 @@ def postprocess_preequilibration(self, data, original_value_dict):
...
+    mapping_scale_opt_to_scale_sim = [scales for _ in range(n_edata)]
+
+    return mapping_scale_opt_to_scale_sim
+
+
+def add_sim_grad_to_opt_grad(par_opt_ids,
+                             mapping_par_opt_to_par_sim,
+                             sim_grad,
+                             opt_grad,
+                             coefficient=-1.0):
```
[the remark this answers is not in the document]
Comment (yannikschaelte) [reply pypesto/objective/amici_objective.py]: yeah, that is the problem with nllh (objective) vs. llh (amici). So I'll have a devault coefficient of 1.0 here, and set to -1.0 in the call.

File: pypesto/objective/petab_import.py [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+            # make list of all timepoints for which measurements exist
+            timepoints = sorted(
+                cur_measurement_df.time.unique().astype(float))
+
+            # init edata object
+            edata = amici.ExpData(self.model.get())
+            edata.setTimepoints(timepoints)
+
+            # handle fixed parameters
+            if len(fixed_parameter_ids) > 0:
```
[the remark this answers is not in the document]
Comment (yannikschaelte) [reply pypesto/objective/petab_import.py]: the hierarchy level continues below the if statement, so continue is not possible here ...

File: pypesto/objective/petab_import.py [hunk 12]
```diff
@@ -0,0 +1,363 @@
...
+                fixed_parameter_vals = condition_df.loc[
+                    condition_df.conditionId ==
```
[the remark this answers is not in the document]
Comment (yannikschaelte) [reply pypesto/objective/petab_import.py]: reset_index() is called, before, so this shoul work, right?

but if reset_index() is not used, would the correct call be 
``fixed_parameter_vals = condition_df[condition.simulationConditionId, fixed_parameter_vals]``?

File: pypesto/objective/petab_import.py (hunk 2 above)
Comment (yannikschaelte) [reply pypesto/objective/petab_import.py]: that's your comment :) . Yes, I also think that we agreed to leave it out now. I'll remove the comment.

File: pypesto/objective/petab_import.py [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+            sigma_y = np.full(
+                shape=(edata.nt(), edata.nytrue()),
+                fill_value=np.nan)
+
+            # add measurements and sigmas
+            for _, measurement in cur_measurement_df.iterrows():
+                time_ix = timepoints.index(measurement.time)
+                observable_ix = observable_ids.index(
+                    f'observable_{measurement.observableId}')
+                # TODO: measurement file should contain prefix
```
Comment (dweindl) [inline pypesto/objective/petab_import.py]:
```suggestion
```

File: pypesto/objective/petab_import.py (hunk 3 above)
Comment (yannikschaelte) [reply pypesto/objective/petab_import.py]: ah of course. should be in parameter_df, if not empty. Yes, I'll make an issue.

**Comment by yannikschaelte** [comment]
> For me this PR is now good to go...

yes, I'm just including the last changes your requested.

**Pull request merged**